// vite-plugin-vue-devtools/dist/client/assets/overview-6ed55b17.js
import { _ as __unplugin_components_0$1 } from './PanelGrids-15f9a979.js';
import { a$ as ref, cc as rpc, a1 as computed, cb as routes, v as defineComponent, r as resolveComponent, o as openBlock, c as createBlock, h as withCtx, x as createBaseVNode, u as createVNode, y as toDisplayString, z as unref, cd as vueVersion, a9 as createTextVNode, e as createElementBlock, F as Fragment, ce as __unplugin_components_0, c3 as __unplugin_components_1, bL as _sfc_main$1 } from './index-55dde797.js';
import { i as isMacOS } from './index-ef144caa.js';

const version = "1.0.0-beta.2";

/* Injected with object hook! */

const modules = ref(await rpc.componentGraph());
const vueModules = computed(() => modules.value.filter((i) => i.id.match(/\.vue($|\?v=)/)));
const pageCount = computed(() => routes.value.length);
const componentCount = computed(() => vueModules.value.length);

/* Injected with object hook! */

const _hoisted_1 = {
  flex: "~ col gap2",
  ma: "",
  "h-full": "",
  "max-w-300": "",
  "w-full": "",
  px20: ""
};
const _hoisted_2 = /* @__PURE__ */ createBaseVNode("div", { "flex-auto": "" }, null, -1);
const _hoisted_3 = {
  flex: "~ col",
  "mt-20": "",
  "items-center": ""
};
const _hoisted_4 = {
  flex: "~",
  "mt--10": "",
  "items-center": "",
  "justify-center": ""
};
const _hoisted_5 = {
  "mr--16": "",
  "mt--6": ""
};
const _hoisted_6 = {
  mb6: "",
  "mt--1": "",
  "text-center": "",
  "text-sm": "",
  flex: "~ gap-1"
};
const _hoisted_7 = /* @__PURE__ */ createBaseVNode("span", { op40: "" }, " Vue DevTools ", -1);
const _hoisted_8 = { op40: "" };
const _hoisted_9 = { flex: "~ gap2 wrap" };
const _hoisted_10 = {
  p4: "",
  "theme-card-green": "",
  flex: "~ col auto"
};
const _hoisted_11 = /* @__PURE__ */ createBaseVNode("div", {
  "logos-vue": "",
  "text-3xl": ""
}, null, -1);
const _hoisted_12 = /* @__PURE__ */ createBaseVNode("div", {
  "carbon-tree-view-alt": "",
  "text-3xl": ""
}, null, -1);
const _hoisted_13 = /* @__PURE__ */ createBaseVNode("div", {
  "i-carbon-assembly-cluster": "",
  "text-3xl": ""
}, null, -1);
const _hoisted_14 = {
  flex: "~ gap-6 wrap",
  "mt-5": "",
  "items-center": "",
  "justify-center": ""
};
const _hoisted_15 = /* @__PURE__ */ createBaseVNode("a", {
  href: "https://github.com/webfansplz/vite-plugin-vue-devtools",
  target: "_blank",
  flex: "~ gap1",
  "items-center": "",
  op50: "",
  hover: "op100 text-blue",
  transition: ""
}, [
  /* @__PURE__ */ createBaseVNode("div", { "i-carbon-star": "" }),
  /* @__PURE__ */ createTextVNode(" Star on GitHub ")
], -1);
const _hoisted_16 = /* @__PURE__ */ createBaseVNode("a", {
  href: "https://github.com/webfansplz/vite-plugin-vue-devtools/discussions/1",
  target: "_blank",
  flex: "~ gap1",
  "items-center": "",
  op50: "",
  hover: "op100 text-yellow",
  transition: ""
}, [
  /* @__PURE__ */ createBaseVNode("div", { "i-carbon-data-enrichment": "" }),
  /* @__PURE__ */ createTextVNode(" Ideas & Suggestions ")
], -1);
const _hoisted_17 = /* @__PURE__ */ createBaseVNode("a", {
  href: "https://github.com/webfansplz/vite-plugin-vue-devtools/discussions/2",
  target: "_blank",
  flex: "~ gap1",
  "items-center": "",
  op50: "",
  hover: "op100 text-lime",
  transition: ""
}, [
  /* @__PURE__ */ createBaseVNode("div", { "i-carbon-plan": "" }),
  /* @__PURE__ */ createTextVNode(" Project Roadmap ")
], -1);
const _hoisted_18 = /* @__PURE__ */ createBaseVNode("a", {
  href: "https://github.com/webfansplz/vite-plugin-vue-devtools/issues",
  target: "_blank",
  flex: "~ gap1",
  "items-center": "",
  op50: "",
  hover: "op100 text-rose",
  transition: ""
}, [
  /* @__PURE__ */ createBaseVNode("div", { "i-carbon-debug": "" }),
  /* @__PURE__ */ createTextVNode(" Bug Reports ")
], -1);
const _hoisted_19 = /* @__PURE__ */ createBaseVNode("div", { "i-carbon-settings": "" }, null, -1);
const _hoisted_20 = /* @__PURE__ */ createBaseVNode("div", { "flex-auto": "" }, null, -1);
const _hoisted_21 = {
  flex: "~ gap-1",
  "cursor-default": "",
  "items-center": "",
  "justify-center": "",
  "pb-8": "",
  "text-sm": "",
  op40: ""
};
const _hoisted_22 = /* @__PURE__ */ createBaseVNode("span", null, "+", -1);
const _hoisted_23 = /* @__PURE__ */ createBaseVNode("span", null, "+", -1);
const _hoisted_24 = /* @__PURE__ */ createBaseVNode("span", null, "+", -1);
const _hoisted_25 = /* @__PURE__ */ createBaseVNode("span", null, "+", -1);
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "overview",
  setup(__props) {
    return (_ctx, _cache) => {
      const _component_Logo = __unplugin_components_0;
      const _component_VDBadge = __unplugin_components_1;
      const _component_RouterLink = resolveComponent("RouterLink");
      const _component_VDButton = _sfc_main$1;
      const _component_VDPanelGrids = __unplugin_components_0$1;
      return openBlock(), createBlock(_component_VDPanelGrids, {
        "h-screen": "",
        "w-full": "",
        flex: "",
        "of-auto": ""
      }, {
        default: withCtx(() => [
          createBaseVNode("div", _hoisted_1, [
            _hoisted_2,
            createBaseVNode("div", _hoisted_3, [
              createBaseVNode("div", _hoisted_4, [
                createVNode(_component_Logo, { "h-18": "" }),
                createBaseVNode("button", _hoisted_5, [
                  createVNode(_component_VDBadge, {
                    "bg-green-400:10": "",
                    "text-green-400": "",
                    title: "preview",
                    textContent: "preview"
                  })
                ])
              ]),
              createBaseVNode("div", _hoisted_6, [
                _hoisted_7,
                createBaseVNode("code", _hoisted_8, "v" + toDisplayString(unref(version)), 1)
              ])
            ]),
            createBaseVNode("div", _hoisted_9, [
              createBaseVNode("div", _hoisted_10, [
                _hoisted_11,
                createBaseVNode("code", null, "v" + toDisplayString(unref(vueVersion)), 1)
              ]),
              createVNode(_component_RouterLink, {
                flex: "~ col auto",
                to: "/pages",
                replace: "",
                "min-w-40": "",
                p4: "",
                "theme-card-lime": ""
              }, {
                default: withCtx(() => [
                  _hoisted_12,
                  createBaseVNode("div", null, toDisplayString(unref(pageCount)) + " pages", 1)
                ]),
                _: 1
              }),
              createVNode(_component_RouterLink, {
                flex: "~ col auto",
                to: "/components",
                replace: "",
                "min-w-40": "",
                p4: "",
                "theme-card-lime": ""
              }, {
                default: withCtx(() => [
                  _hoisted_13,
                  createBaseVNode("div", null, toDisplayString(unref(componentCount)) + " components", 1)
                ]),
                _: 1
              })
            ]),
            createBaseVNode("div", _hoisted_14, [
              _hoisted_15,
              _hoisted_16,
              _hoisted_17,
              _hoisted_18,
              createVNode(_component_RouterLink, {
                to: "/settings",
                flex: "~ gap1",
                replace: "",
                "inline-block": "",
                "items-center": "",
                op50: "",
                "hover:op80": ""
              }, {
                default: withCtx(() => [
                  _hoisted_19,
                  createTextVNode(" Settings ")
                ]),
                _: 1
              })
            ]),
            _hoisted_20,
            createBaseVNode("div", _hoisted_21, [
              createTextVNode(" Press "),
              unref(isMacOS)() ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
                createVNode(_component_VDButton, {
                  n: "xs",
                  class: "px2"
                }, {
                  default: withCtx(() => [
                    createTextVNode(" ⇧ Shift ")
                  ]),
                  _: 1
                }),
                _hoisted_22,
                createVNode(_component_VDButton, {
                  n: "xs",
                  class: "px2"
                }, {
                  default: withCtx(() => [
                    createTextVNode(" ⌥ Option ")
                  ]),
                  _: 1
                }),
                _hoisted_23,
                createVNode(_component_VDButton, {
                  n: "xs",
                  class: "px2"
                }, {
                  default: withCtx(() => [
                    createTextVNode(" D ")
                  ]),
                  _: 1
                })
              ], 64)) : (openBlock(), createElementBlock(Fragment, { key: 1 }, [
                createVNode(_component_VDButton, {
                  n: "xs",
                  class: "px2"
                }, {
                  default: withCtx(() => [
                    createTextVNode(" Shift ")
                  ]),
                  _: 1
                }),
                _hoisted_24,
                createVNode(_component_VDButton, {
                  n: "xs",
                  class: "px2"
                }, {
                  default: withCtx(() => [
                    createTextVNode(" Alt ")
                  ]),
                  _: 1
                }),
                _hoisted_25,
                createVNode(_component_VDButton, {
                  n: "xs",
                  class: "px2"
                }, {
                  default: withCtx(() => [
                    createTextVNode(" D ")
                  ]),
                  _: 1
                })
              ], 64)),
              createTextVNode(" to toggle DevTools ")
            ])
          ])
        ]),
        _: 1
      });
    };
  }
});

/* Injected with object hook! */

/* Injected with object hook! */

export { _sfc_main as default };
