// vite-plugin-vue-devtools/dist/client/assets/pinia-99843507.js
import { _ as _sfc_main$1 } from './StateFields.vue_vue_type_script_setup_true_lang-eaba3733.js';
import { g, M } from './splitpanes.es-ca653775.js';
import { v as defineComponent, a$ as ref, a1 as computed, c7 as piniaStoresCategory, bd as toRaw, c8 as piniaState, c9 as piniaGetters, z as unref, o as openBlock, e as createElementBlock, u as createVNode, h as withCtx, x as createBaseVNode, F as Fragment, i as renderList, q as normalizeClass, y as toDisplayString, c as createBlock } from './index-55dde797.js';
import { p as pick } from './index-ef144caa.js';
import './Icon.vue_vue_type_script_setup_true_lang-5b8e75b9.js';
import './IconButton.vue_vue_type_script_setup_true_lang-1af9c374.js';

const _hoisted_1 = {
  "h-screen": "",
  "n-panel-grids": ""
};
const _hoisted_2 = {
  "h-screen": "",
  "select-none": "",
  "overflow-scroll": "",
  "p-2": "",
  class: "no-scrollbar"
};
const _hoisted_3 = ["onClick"];
const _hoisted_4 = { "vue-block-title": "" };
const _hoisted_5 = {
  "h-screen": "",
  "select-none": "",
  "overflow-scroll": "",
  "p-2": "",
  class: "no-scrollbar"
};
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "pinia",
  setup(__props) {
    const activeIndex = ref(0);
    const pickStoreId = computed(() => {
      return activeIndex.value === 0 ? "" : piniaStoresCategory.value[activeIndex.value];
    });
    function select(index) {
      activeIndex.value = index;
    }
    const data = computed(() => {
      const state = pickStoreId.value ? pick(toRaw(piniaState.value), [pickStoreId.value]) : piniaState.value;
      const getters = pickStoreId.value ? pick(toRaw(piniaGetters.value), [pickStoreId.value]) : piniaGetters.value;
      const hasState = Object.values(unref(state)).some((item) => item !== void 0);
      const hasGetters = Object.values(unref(getters)).some((item) => item !== void 0);
      return [
        hasState && {
          key: "state",
          value: unref(state)
        },
        hasGetters && {
          key: "getters",
          value: unref(getters)
        }
      ].filter(Boolean);
    });
    return (_ctx, _cache) => {
      const _component_StateFields = _sfc_main$1;
      return openBlock(), createElementBlock("div", _hoisted_1, [
        createVNode(unref(M), null, {
          default: withCtx(() => [
            createVNode(unref(g), {
              border: "r base",
              size: 40
            }, {
              default: withCtx(() => [
                createBaseVNode("div", _hoisted_2, [
                  (openBlock(true), createElementBlock(Fragment, null, renderList(unref(piniaStoresCategory), (item, index) => {
                    return openBlock(), createElementBlock("div", {
                      key: index,
                      "vue-block": "",
                      class: normalizeClass([unref(activeIndex) === index ? "vue-block-active" : "vue-block-hover"]),
                      onClick: ($event) => select(index)
                    }, [
                      createBaseVNode("h3", _hoisted_4, [
                        createBaseVNode("span", {
                          class: normalizeClass([unref(activeIndex) === index && "text-white"])
                        }, toDisplayString(item), 3)
                      ])
                    ], 10, _hoisted_3);
                  }), 128))
                ])
              ]),
              _: 1
            }),
            createVNode(unref(g), { size: 60 }, {
              default: withCtx(() => [
                createBaseVNode("div", _hoisted_5, [
                  (openBlock(true), createElementBlock(Fragment, null, renderList(unref(data), (item, index) => {
                    return openBlock(), createBlock(_component_StateFields, {
                      id: index,
                      key: index,
                      data: item
                    }, null, 8, ["id", "data"]);
                  }), 128))
                ])
              ]),
              _: 1
            })
          ]),
          _: 1
        })
      ]);
    };
  }
});

/* Injected with object hook! */

/* Injected with object hook! */

export { _sfc_main as default };
