// vite-plugin-vue-devtools/dist/client/assets/rerender-trace-a1fc81e7.js
import { v as defineComponent, a$ as ref, c4 as DevToolsHooks, r as resolveComponent, o as openBlock, e as createElementBlock, x as createBaseVNode, u as createVNode, h as withCtx, z as unref, c as createBlock, a9 as createTextVNode, g as createCommentVNode, F as Fragment, i as renderList, c5 as useDevToolsClient, c6 as rootPath, _ as __unplugin_components_7, c3 as __unplugin_components_1, p as pushScopeId, a as popScopeId, y as toDisplayString, bJ as _export_sfc } from './index-55dde797.js';
import { d as dayjs } from './dayjs.min-3c1187ae.js';
import { g as getSetupStateInfo, t as toRaw } from './data-1539e705.js';
import './_commonjsHelpers-d60b99f1.js';

const _withScopeId = (n) => (pushScopeId("data-v-c4e9e7b3"), n = n(), popScopeId(), n);
const _hoisted_1 = {
  "h-full": "",
  "w-full": "",
  "of-auto": "",
  "n-panel-grids": ""
};
const _hoisted_2 = { "w-full": "" };
const _hoisted_3 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createBaseVNode("p", {
  "text-xs": "",
  "op-50": ""
}, " Start Tracing ", -1));
const _hoisted_4 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createBaseVNode("p", {
  "text-xs": "",
  "op-50": ""
}, " Clear ", -1));
const _hoisted_5 = {
  key: 0,
  flex: "~ items-center justify-center",
  "h-screen": ""
};
const _hoisted_6 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createBaseVNode("i", {
  class: "i-fluent:record-12-regular",
  text: "4.7",
  "cursor-pointer": "",
  "text-secondary": "",
  hover: "text-black dark:text-white"
}, null, -1));
const _hoisted_7 = [
  _hoisted_6
];
const _hoisted_8 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createBaseVNode("p", { flex: "~ items-center" }, [
  /* @__PURE__ */ createTextVNode(" Click the clear button "),
  /* @__PURE__ */ createBaseVNode("span", {
    border: "~ base",
    "mx-1": "",
    "inline-flex": "",
    "items-center": "",
    "justify-center": "",
    p1: ""
  }, [
    /* @__PURE__ */ createBaseVNode("i", {
      class: "i-grommet-icons:clear",
      text: "3.8",
      "text-secondary": ""
    })
  ]),
  /* @__PURE__ */ createTextVNode(" to clear the results. ")
], -1));
const _hoisted_9 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createBaseVNode("p", { "lh-5": "" }, [
  /* @__PURE__ */ createTextVNode(" Start tracing, do something in the page. Then,"),
  /* @__PURE__ */ createBaseVNode("br"),
  /* @__PURE__ */ createTextVNode(" press the stop button to show the resulting triggers for page rerenders. ")
], -1));
const _hoisted_10 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createBaseVNode("p", { class: "loading-animation" }, " Tracing... ", -1));
const _hoisted_11 = { key: 1 };
const _hoisted_12 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createBaseVNode("thead", {
  border: "b r base",
  sticky: "",
  display: "table-header-group",
  "top-0": "",
  "z-10": ""
}, [
  /* @__PURE__ */ createBaseVNode("tr", { "bg-base": "" }, [
    /* @__PURE__ */ createBaseVNode("th", {
      "p-2": "",
      "text-center": ""
    }, " Component File "),
    /* @__PURE__ */ createBaseVNode("th", {
      "p-2": "",
      "text-center": ""
    }, " Key "),
    /* @__PURE__ */ createBaseVNode("th", {
      "p-2": "",
      "text-center": ""
    }, " Value "),
    /* @__PURE__ */ createBaseVNode("th", {
      "p-2": "",
      "text-center": ""
    }, " Data Type "),
    /* @__PURE__ */ createBaseVNode("th", {
      "p-2": "",
      "text-center": ""
    }, " Update Type "),
    /* @__PURE__ */ createBaseVNode("th", {
      "p-2": "",
      "text-center": ""
    }, " Update Time ")
  ])
], -1));
const _hoisted_13 = { border: "b r base" };
const _hoisted_14 = ["onClick"];
const _hoisted_15 = {
  "max-w-30": "",
  "w-30": "",
  "of-hidden": "",
  "text-ellipsis": "",
  "ws-nowrap": "",
  "px-2": "",
  "text-center": "",
  "text-sm": "",
  "font-mono": "",
  op70: ""
};
const _hoisted_16 = {
  "max-w-30": "",
  "w-30": "",
  "of-hidden": "",
  "text-ellipsis": "",
  "ws-nowrap": "",
  "px-2": "",
  "text-center": "",
  "text-sm": "",
  "font-mono": "",
  op70: ""
};
const _hoisted_17 = {
  "max-w-30": "",
  "w-30": "",
  "of-hidden": "",
  "text-ellipsis": "",
  "ws-nowrap": "",
  "px-2": "",
  "text-center": "",
  "text-sm": "",
  "font-mono": "",
  op70: ""
};
const _hoisted_18 = {
  "max-w-30": "",
  "w-30": "",
  "of-hidden": "",
  "text-ellipsis": "",
  "ws-nowrap": "",
  "px-2": "",
  "text-center": "",
  "text-sm": "",
  "font-mono": "",
  op70: ""
};
const _hoisted_19 = {
  "max-w-30": "",
  "w-30": "",
  "of-hidden": "",
  "text-ellipsis": "",
  "ws-nowrap": "",
  "px-2": "",
  "text-center": "",
  "text-sm": "",
  "font-mono": "",
  op70: ""
};
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "rerender-trace",
  setup(__props) {
    const client = useDevToolsClient();
    const hook = client.value.hook;
    const isTracing = ref(false);
    const result = ref([]);
    const traceBuffer = ref([]);
    function normalizeEventInfo(e, instance) {
      const info = getSetupStateInfo(e.target);
      const file = instance.type.__file?.replace?.(rootPath, "") ?? "-";
      const dataType = info.computed ? "Computed" : info.ref ? "Ref" : info.reactive ? "Reactive" : null;
      const index = Object.values(instance.devtoolsRawSetupState).map((i) => toRaw(i)).indexOf(toRaw(e.target));
      const key = index === -1 ? e.key ?? "unknown" : Object.keys(instance.devtoolsRawSetupState)[index];
      const value = !dataType || info.reactive ? e.target[e.key] : e.target.value;
      return {
        componentFile: file,
        fullFilePath: instance.type.__file ?? "",
        key,
        value,
        dataType: dataType ?? "unknown",
        updateType: e.type ?? "-",
        updateTime: Date.now()
      };
    }
    hook.on(DevToolsHooks.RENDER_TRACKED, (e, instance) => {
      isTracing.value && traceBuffer.value.push(normalizeEventInfo(e, instance));
    });
    hook.on(DevToolsHooks.RENDER_TRIGGERED, (e, instance) => {
      isTracing.value && traceBuffer.value.push(normalizeEventInfo(e, instance));
    });
    function start() {
      result.value = [];
      traceBuffer.value = [];
      isTracing.value = true;
    }
    function stop() {
      isTracing.value = false;
      const typeOrder = { Ref: 0, Computed: 1 };
      result.value = traceBuffer.value.sort((a, b) => typeOrder[a.dataType] - typeOrder[b.dataType]).sort((a, b) => a.updateTime - b.updateTime);
      traceBuffer.value = [];
    }
    function clear() {
      if (isTracing.value)
        return;
      result.value = [];
    }
    function openInEditor(filePath) {
      if (!filePath)
        return;
      client.value.openInEditor(filePath);
    }
    return (_ctx, _cache) => {
      const _component_VTooltip = resolveComponent("VTooltip");
      const _component_VDCard = __unplugin_components_7;
      const _component_VDBadge = __unplugin_components_1;
      return openBlock(), createElementBlock("div", _hoisted_1, [
        createBaseVNode("div", _hoisted_2, [
          createVNode(_component_VDCard, {
            flex: "~ col gap-2",
            absolute: "",
            "right-3": "",
            "top-3": "",
            "z-20": "",
            p1: ""
          }, {
            default: withCtx(() => [
              createVNode(_component_VTooltip, {
                placement: "bottom",
                distance: 12,
                "text-center": ""
              }, {
                popper: withCtx(() => [
                  _hoisted_3
                ]),
                default: withCtx(() => [
                  !unref(isTracing) ? (openBlock(), createElementBlock("i", {
                    key: 0,
                    class: "i-fluent:record-12-regular",
                    text: "4.7",
                    "cursor-pointer": "",
                    "text-secondary": "",
                    hover: "text-black dark:text-white",
                    onClick: start
                  })) : (openBlock(), createElementBlock("i", {
                    key: 1,
                    class: "i-fluent:record-stop-12-regular",
                    text: "4.7",
                    "cursor-pointer": "",
                    "text-red-600": "",
                    hover: "text-red-800",
                    onClick: stop
                  }))
                ]),
                _: 1
              }),
              createVNode(_component_VTooltip, {
                placement: "bottom",
                distance: 12,
                "text-center": ""
              }, {
                popper: withCtx(() => [
                  _hoisted_4
                ]),
                default: withCtx(() => [
                  createBaseVNode("i", {
                    class: "i-grommet-icons:clear",
                    text: "3.8",
                    "cursor-pointer": "",
                    "text-secondary": "",
                    hover: "text-black dark:text-white",
                    onClick: clear
                  })
                ]),
                _: 1
              })
            ]),
            _: 1
          }),
          !unref(result).length ? (openBlock(), createElementBlock("div", _hoisted_5, [
            !unref(isTracing) && !unref(result).length ? (openBlock(), createBlock(_component_VDCard, {
              key: 0,
              flex: "~ col gap2",
              relative: "",
              p3: "",
              op60: ""
            }, {
              default: withCtx(() => [
                createBaseVNode("p", { flex: "~ items-center" }, [
                  createTextVNode(" Click the start button "),
                  createBaseVNode("button", {
                    border: "~ base",
                    "mx-1": "",
                    "inline-flex": "",
                    "items-center": "",
                    "justify-center": "",
                    p: "0.5",
                    onClick: start
                  }, _hoisted_7),
                  createTextVNode(" to start a new tracing. ")
                ]),
                _hoisted_8,
                _hoisted_9,
                createVNode(_component_VDBadge, {
                  absolute: "",
                  "right--0.5": "",
                  "top-0": "",
                  "rounded-0": "",
                  "bg-green-400:10": "",
                  "text-green-400": "",
                  title: "Experimental",
                  textContent: "Experimental"
                })
              ]),
              _: 1
            })) : unref(isTracing) && !unref(result).length ? (openBlock(), createBlock(_component_VDCard, {
              key: 1,
              flex: "~ col gap2",
              p3: ""
            }, {
              default: withCtx(() => [
                _hoisted_10
              ]),
              _: 1
            })) : createCommentVNode("", true)
          ])) : createCommentVNode("", true),
          unref(result).length && !unref(isTracing) ? (openBlock(), createElementBlock("table", _hoisted_11, [
            _hoisted_12,
            createBaseVNode("tbody", _hoisted_13, [
              (openBlock(true), createElementBlock(Fragment, null, renderList(unref(result), (item, index) => {
                return openBlock(), createElementBlock("tr", {
                  key: index,
                  class: "group",
                  "h-7": "",
                  border: "b dashed transparent hover:base"
                }, [
                  createBaseVNode("td", {
                    "max-w-50": "",
                    "of-hidden": "",
                    "text-ellipsis": "",
                    "px-2": "",
                    "text-center": "",
                    "text-sm": "",
                    underline: "",
                    op70: "",
                    hover: "text-primary",
                    onClick: ($event) => openInEditor(item.fullFilePath)
                  }, toDisplayString(item.componentFile), 9, _hoisted_14),
                  createBaseVNode("td", _hoisted_15, toDisplayString(item.key), 1),
                  createBaseVNode("td", _hoisted_16, toDisplayString(item.value), 1),
                  createBaseVNode("td", _hoisted_17, toDisplayString(item.dataType), 1),
                  createBaseVNode("td", _hoisted_18, toDisplayString(item.updateType), 1),
                  createBaseVNode("td", _hoisted_19, toDisplayString(unref(dayjs)(item.updateTime).format("HH:mm:ss")), 1)
                ]);
              }), 128))
            ])
          ])) : createCommentVNode("", true)
        ])
      ]);
    };
  }
});

/* Injected with object hook! */

/* unplugin-vue-components disabled */const rerenderTrace_vue_vue_type_style_index_0_scoped_c4e9e7b3_lang = '';
/* Injected with object hook! */

const rerenderTrace = /* @__PURE__ */ _export_sfc(_sfc_main, [["__scopeId", "data-v-c4e9e7b3"]]);

/* Injected with object hook! */

export { rerenderTrace as default };
