// vite-plugin-vue-devtools/dist/client/assets/routes-56bbe197.js
import { _ as __unplugin_components_0 } from './PanelGrids-15f9a979.js';
import { v as defineComponent, z as unref, b_ as router, o as openBlock, e as createElementBlock, u as createVNode, h as withCtx, x as createBaseVNode, F as Fragment, i as renderList, b$ as routeRecordMatcherState, c0 as activeRouteRecordMatcherState, c as createBlock, bZ as _sfc_main$2, _ as __unplugin_components_7, q as normalizeClass, c1 as activeRouteRecordIndex, c2 as toggleRouteRecordMatcher, a9 as createTextVNode, y as toDisplayString, l as normalizeStyle, g as createCommentVNode, c3 as __unplugin_components_1 } from './index-55dde797.js';
import { _ as _sfc_main$1 } from './StateFields.vue_vue_type_script_setup_true_lang-eaba3733.js';
import { g, M } from './splitpanes.es-ca653775.js';
import './Icon.vue_vue_type_script_setup_true_lang-5b8e75b9.js';
import './index-ef144caa.js';
import './IconButton.vue_vue_type_script_setup_true_lang-1af9c374.js';

const _hoisted_1 = {
  key: 0,
  "h-screen": "",
  "n-panel-grids": ""
};
const _hoisted_2 = {
  "h-screen": "",
  "select-none": "",
  "overflow-scroll": "",
  "p-2": "",
  class: "no-scrollbar"
};
const _hoisted_3 = ["onClick"];
const _hoisted_4 = { "vue-block-title": "" };
const _hoisted_5 = {
  "h-screen": "",
  "select-none": "",
  "overflow-scroll": "",
  "p-2": "",
  class: "no-scrollbar"
};
const _hoisted_6 = /* @__PURE__ */ createBaseVNode("h1", { "text-xl": "" }, " Install Vue Router ", -1);
const _hoisted_7 = /* @__PURE__ */ createBaseVNode("p", {
  "text-sm": "",
  op50: ""
}, " It seems you don't have vue-router installed. ", -1);
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "routes",
  setup(__props) {
    return (_ctx, _cache) => {
      const _component_VDBadge = __unplugin_components_1;
      const _component_StateFields = _sfc_main$1;
      const _component_TabIcon = _sfc_main$2;
      const _component_VDCard = __unplugin_components_7;
      const _component_VDPanelGrids = __unplugin_components_0;
      return unref(router) ? (openBlock(), createElementBlock("div", _hoisted_1, [
        createVNode(unref(M), null, {
          default: withCtx(() => [
            createVNode(unref(g), {
              border: "r base",
              size: "40"
            }, {
              default: withCtx(() => [
                createBaseVNode("div", _hoisted_2, [
                  (openBlock(true), createElementBlock(Fragment, null, renderList(unref(routeRecordMatcherState), (item, index) => {
                    return openBlock(), createElementBlock("div", {
                      key: index,
                      "vue-block": "",
                      class: normalizeClass([unref(activeRouteRecordIndex) === index ? "vue-block-active" : "vue-block-hover"]),
                      onClick: ($event) => unref(toggleRouteRecordMatcher)(index)
                    }, [
                      createBaseVNode("h3", _hoisted_4, [
                        createBaseVNode("span", {
                          class: normalizeClass([unref(activeRouteRecordIndex) === index && "text-white"])
                        }, [
                          createTextVNode(toDisplayString(item.path) + " ", 1),
                          (openBlock(true), createElementBlock(Fragment, null, renderList(item.tags, (tag, childIndex) => {
                            return openBlock(), createElementBlock(Fragment, null, [
                              tag.label ? (openBlock(), createBlock(_component_VDBadge, {
                                key: childIndex,
                                color: "white/90 dark:black/90",
                                style: normalizeStyle({
                                  backgroundColor: `#${tag.bgColor.toString(16)}`
                                })
                              }, {
                                default: withCtx(() => [
                                  createTextVNode(toDisplayString(tag.label), 1)
                                ]),
                                _: 2
                              }, 1032, ["style"])) : createCommentVNode("", true)
                            ], 64);
                          }), 256))
                        ], 2)
                      ])
                    ], 10, _hoisted_3);
                  }), 128))
                ])
              ]),
              _: 1
            }),
            createVNode(unref(g), { size: "60" }, {
              default: withCtx(() => [
                createBaseVNode("div", _hoisted_5, [
                  createVNode(_component_StateFields, { data: unref(activeRouteRecordMatcherState) }, null, 8, ["data"])
                ])
              ]),
              _: 1
            })
          ]),
          _: 1
        })
      ])) : (openBlock(), createBlock(_component_VDPanelGrids, {
        key: 1,
        px5: ""
      }, {
        default: withCtx(() => [
          createVNode(_component_VDCard, {
            flex: "~ col gap2",
            "min-w-30": "",
            "items-center": "",
            p3: ""
          }, {
            default: withCtx(() => [
              createVNode(_component_TabIcon, {
                mb2: "",
                "text-5xl": "",
                icon: "i-logos-vue",
                title: "Vue Router"
              }),
              _hoisted_6,
              _hoisted_7
            ]),
            _: 1
          })
        ]),
        _: 1
      }));
    };
  }
});

/* Injected with object hook! */

/* Injected with object hook! */

export { _sfc_main as default };
