// vite-plugin-vue-devtools/dist/client/assets/SectionBlock-22f38034.js
import { _ as _sfc_main$1 } from './IconTitle.vue_vue_type_script_setup_true_lang-4fc5b6ac.js';
import { _ as _sfc_main$2 } from './Icon.vue_vue_type_script_setup_true_lang-5b8e75b9.js';
import { v as defineComponent, bK as useVModel, b as resolveDirective, o as openBlock, e as createElementBlock, x as createBaseVNode, q as normalizeClass, u as createVNode, h as withCtx, f as renderSlot, g as createCommentVNode, c as createBlock, z as unref, d as withDirectives, F as Fragment, a9 as createTextVNode, y as toDisplayString, p as pushScopeId, a as popScopeId, bJ as _export_sfc } from './index-55dde797.js';

const _withScopeId = (n) => (pushScopeId("data-v-e966339e"), n = n(), popScopeId(), n);
const _hoisted_1 = ["open"];
const _hoisted_2 = { "text-base": "" };
const _hoisted_3 = {
  key: 0,
  "text-sm": "",
  op50: ""
};
const _hoisted_4 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createBaseVNode("div", { class: "flex-auto" }, null, -1));
const _hoisted_5 = /* @__PURE__ */ _withScopeId(() => /* @__PURE__ */ createBaseVNode("div", { class: "x-divider" }, null, -1));
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "SectionBlock",
  props: {
    icon: {},
    text: {},
    description: {},
    containerClass: { default: "" },
    collapse: { type: Boolean, default: true },
    open: { type: Boolean, default: true },
    padding: { type: [Boolean, String], default: true }
  },
  setup(__props) {
    const props = __props;
    const open = useVModel(props, "open", void 0, { passive: true });
    function onToggle(e) {
      open.value = e.target.open;
    }
    return (_ctx, _cache) => {
      const _component_VDIcon = _sfc_main$2;
      const _component_VDIconTitle = _sfc_main$1;
      const _directive_lazy_show = resolveDirective("lazy-show");
      return openBlock(), createElementBlock(Fragment, null, [
        createBaseVNode("details", {
          open: unref(open),
          onToggle
        }, [
          createBaseVNode("summary", {
            class: normalizeClass(["cursor-pointer select-none hover:bg-active p4", _ctx.collapse ? "" : "pointer-events-none"])
          }, [
            createVNode(_component_VDIconTitle, {
              icon: _ctx.icon,
              text: _ctx.text,
              "text-xl": "",
              transition: "",
              class: normalizeClass(unref(open) ? "op100" : "op60")
            }, {
              default: withCtx(() => [
                createBaseVNode("div", null, [
                  createBaseVNode("div", _hoisted_2, [
                    renderSlot(_ctx.$slots, "text", {}, () => [
                      createTextVNode(toDisplayString(_ctx.text), 1)
                    ], true)
                  ]),
                  _ctx.description || _ctx.$slots.description ? (openBlock(), createElementBlock("div", _hoisted_3, [
                    renderSlot(_ctx.$slots, "description", {}, () => [
                      createTextVNode(toDisplayString(_ctx.description), 1)
                    ], true)
                  ])) : createCommentVNode("", true)
                ]),
                _hoisted_4,
                renderSlot(_ctx.$slots, "actions", {}, void 0, true),
                _ctx.collapse ? (openBlock(), createBlock(_component_VDIcon, {
                  key: 0,
                  icon: "carbon-chevron-down",
                  class: "chevron",
                  "cursor-pointer": "",
                  "place-self-start": "",
                  "text-base": "",
                  op75: "",
                  transition: "",
                  "duration-500": ""
                })) : createCommentVNode("", true)
              ]),
              _: 3
            }, 8, ["icon", "text", "class"])
          ], 2),
          withDirectives((openBlock(), createElementBlock("div", {
            class: normalizeClass(["flex flex-col flex-gap2 pb6 pt2", typeof _ctx.padding === "string" ? _ctx.padding : _ctx.padding ? "px4" : ""])
          }, [
            renderSlot(_ctx.$slots, "details", {}, void 0, true),
            createBaseVNode("div", {
              class: normalizeClass([_ctx.containerClass, "mt1"])
            }, [
              renderSlot(_ctx.$slots, "default", {}, void 0, true)
            ], 2),
            renderSlot(_ctx.$slots, "footer", {}, void 0, true)
          ], 2)), [
            [_directive_lazy_show, unref(open)]
          ])
        ], 40, _hoisted_1),
        _hoisted_5
      ], 64);
    };
  }
});

/* Injected with object hook! */

/* unplugin-vue-components disabled */const SectionBlock_vue_vue_type_style_index_0_scoped_e966339e_lang = '';
/* Injected with object hook! */

const __unplugin_components_3 = /* @__PURE__ */ _export_sfc(_sfc_main, [["__scopeId", "data-v-e966339e"]]);

/* Injected with object hook! */

export { __unplugin_components_3 as _ };
